File: javascript/2. Add Two Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let answer = new ListNode(); // 答案
    let tempLinkedList = answer; // 用來暫存每一個答案的linked list
    let jingWei = 0; // 進位紀錄
    // 條件為: 只要l1, l2有一個還有連結下個linked list或是有進位
    while(l1 !== null || l2 !== null || jingWei != 0) {
        const l1Val = l1 ? l1.val : 0;
        const l2Val = l2 ? l2.val : 0;
        let sum = (l1Val + l2Val + jingWei);
        let residual = 0; // 加總扣除進位後的餘數, 代表每一個linked list的value
        if(sum >= 10) {
            jingWei = 1;
            residual = sum - 10;
        } 
        else {
            jingWei = 0;
            residual = sum;
        }
        
        tempLinkedList.val = residual;
        // 指向l1, l2的下一個 linked list
        l1 = l1 ? l1.next : null; 
        l2 = l2 ? l2.next : null;
        // 避免創建下一個空的linked list的方法，目前沒想到更好的做法
        if(l1 === null && l2 === null && jingWei === 0) break;
        // 創建並指向下一個答案的linked list
        tempLinkedList.next = new ListNode();
        tempLinkedList = tempLinkedList.next;
    }
    return answer;
};

File: java/2. Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int carry = 0;
    public ListNode dummy = null;
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        dummy = l1;
        while (l1 != null || l2 != null) {
            l1.val = calculation(l1, l2);
            if (l1.next == null && l2.next == null) break;
            else if (l1.next == null && l2.next != null) l1.next = new ListNode(0);
            else if (l1.next != null && l2.next == null) l2.next = new ListNode(0);
            l1 = l1.next;
            l2 = l2.next;
        }

        if (carry > 0) l1.next = new ListNode(carry);

        return dummy;
    }

    public int calculation(ListNode l1, ListNode l2) {
        boolean isCarry = (l1.val + l2.val + carry) >= 10;
        int val = isCarry ? (l1.val + l2.val + carry - 10) : (l1.val + l2.val + carry);
        carry = isCarry ? 1 : 0;
        return val;
    }
}